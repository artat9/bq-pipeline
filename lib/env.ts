const projectName = "kaleido-backend";

export enum Environments {
  PROD = "prod",
  DEV = "dev",
  TEST = "test",
}
export interface EnvironmentVariables {
  rootDomain: string;
  hostedZoneId: string;
  txtRecords?: string[];
  allowedOrigin: string;
  deployBranchName: string;
  certificateArn: string;
}

const EnvironmentVariablesSetting: {
  [key in Environments]: EnvironmentVariables;
} = {
  [Environments.PROD]: {
    certificateArn:
      "arn:aws:acm:us-east-1:495476032358:certificate/98c65aba-1677-404e-9bac-1b2142602ad2",
    rootDomain: "kaleidodao.org",
    hostedZoneId: "Z07302833NYNT52C5SA30",
    allowedOrigin: "https://kaleidodao.org",
    deployBranchName: "main",
  },
  [Environments.DEV]: {
    certificateArn:
      "arn:aws:acm:us-east-1:495476032358:certificate/c635e3e6-7cc5-4777-b0a9-57317a70e9c2",
    rootDomain: "kaleido-dev.tk",
    hostedZoneId: "Z04820191RSTGGR5Z7MK0",
    allowedOrigin: "https://kaleido-webfront-git-dev-squard.vercel.app",
    deployBranchName: "dev",
  },
  [Environments.TEST]: {
    certificateArn: "",
    rootDomain: "test.example.com",
    hostedZoneId: "XXX",
    allowedOrigin: "http://localhost:3001",
    deployBranchName: "test",
  },
};

export function valueOf(env: Environments): EnvironmentVariables {
  return EnvironmentVariablesSetting[env];
}
export const withEnvPrefix = (target: Environments, str: string) =>
  `${projectName}-${str}-${target}`;

export const isProd = (target: Environments) => {
  return target === Environments.PROD;
};
export const isDev = (target: Environments) => {
  return target === Environments.DEV;
};
