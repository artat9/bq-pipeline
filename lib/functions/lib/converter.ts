import svgToImg from "@lildiary/svg-to-img";
import { ConvertGltfToGLB } from "gltf-import-export";
import chronium = require("chrome-aws-lambda");

import path = require("path");

export const toGLB = (gltbFileName: string, outputFileName: string): void => {
  ConvertGltfToGLB(pathFromAssetName(gltbFileName), outputFileName);
};

const pathFromAssetName = (filename: string) => {
  return path.join(__dirname, "..", "voucher", "assets", filename);
};

export const toJpeg = async (svg: string, dist: string) => {
  let browser = null;
  await svgToImg
    .from(svg)
    .toJpeg({
      path: dist,
    })
    .catch((e) => {
      console.log(e);
    })
    .then((t) => {
      console.log(t);
    })
    .finally(() => console.log("finished"));
};
