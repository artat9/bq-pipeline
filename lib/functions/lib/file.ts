import path = require("path");
import * as fs from "fs";

export const fromSVGToString = (filename: string): string => {
  const buffer = fs.readFileSync(path.join(__dirname, "assets", filename));
  return buffer.toString();
};
