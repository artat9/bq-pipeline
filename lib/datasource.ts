import * as dynamodb from "@aws-cdk/aws-dynamodb";
import * as cdk from "@aws-cdk/core";
import * as environment from "./env";

export class DataSourceStack extends cdk.Stack {
  public readonly ddb: dynamodb.Table;
  constructor(
    scope: cdk.Construct,
    id: string,
    target: environment.Environments
  ) {
    super(scope, id);
    const tableName = environment.withEnvPrefix(target, "main");
    this.ddb = new dynamodb.Table(this, tableName, ddbProps(tableName));
    withIndexes(this.ddb);
  }
}

const withIndexes = (table: dynamodb.Table) => {
  table.addGlobalSecondaryIndex({
    indexName: "GSI-1",
    partitionKey: {
      name: "SK",
      type: dynamodb.AttributeType.STRING,
    },
    sortKey: {
      name: "Data",
      type: dynamodb.AttributeType.STRING,
    },
  });
  table.addGlobalSecondaryIndex({
    indexName: "GSI-2",
    partitionKey: {
      name: "PK",
      type: dynamodb.AttributeType.STRING,
    },
    sortKey: {
      name: "Number",
      type: dynamodb.AttributeType.NUMBER,
    },
  });
};

const ddbProps = (tableName: string): dynamodb.TableProps => {
  return {
    tableName: tableName,
    partitionKey: {
      name: `PK`,
      type: dynamodb.AttributeType.STRING,
    },
    sortKey: {
      name: `SK`,
      type: dynamodb.AttributeType.STRING,
    },
    billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
    removalPolicy: cdk.RemovalPolicy.RETAIN,
    pointInTimeRecovery: true,
  };
};
