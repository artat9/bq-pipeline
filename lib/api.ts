import {
  AuthorizationType,
  DynamoDbDataSource,
  FieldLogLevel,
  GraphqlApi,
  GraphqlApiProps,
  LambdaDataSource,
  MappingTemplate,
  Schema,
} from "@aws-cdk/aws-appsync";
import * as dynamodb from "@aws-cdk/aws-dynamodb";
import * as cdk from "@aws-cdk/core";
import { join } from "path";
import * as environment from "./env";
import { lambdaFunction } from "./function";
import IAM = require("@aws-cdk/aws-iam");
interface ApiProps extends cdk.StackProps {
  ddb: dynamodb.Table;
}

export class ApiStack extends cdk.Stack {
  public readonly appSyncLog: string;
  constructor(
    scope: cdk.Construct,
    id: string,
    target: environment.Environments,
    props: ApiProps
  ) {
    super(scope, id, props);
    const api = new GraphqlApi(this, "GraphqlApi", apiProps(target));
    withDDBResolvers(api, dDBDataSource(api, target, props));
    api.createResolver({
      typeName: "Mutation",
      fieldName: "uploadImage",
      dataSource: new LambdaDataSource(this, "uploadimage", {
        api: api,
        lambdaFunction: lambdaFunction(this, "issuepresign", target),
      }),
    });
    api.createResolver({
      typeName: "Mutation",
      fieldName: "applyForMediaAccount",
      dataSource: new LambdaDataSource(this, "createmediaaccount", {
        api: api,
        lambdaFunction: lambdaFunction(this, "mediaaccount", target),
      }),
    });
  }
}

const withDDBResolvers = (
  api: GraphqlApi,
  ddbDataSource: DynamoDbDataSource
) => {
  api.createResolver({
    typeName: "Query",
    fieldName: "getAccount",
    dataSource: ddbDataSource,
    requestMappingTemplate: MappingTemplate.fromFile(
      join(
        __dirname,
        "schema",
        "resolvers",
        "public",
        "getAccount" + ".request.vtl"
      )
    ),
    responseMappingTemplate: MappingTemplate.fromFile(
      join(
        __dirname,
        "schema",
        "resolvers",
        "public",
        "getAccount" + ".response.vtl"
      )
    ),
  });
};

const dDBDataSource = (
  api: GraphqlApi,
  target: environment.Environments,
  props: ApiProps
) => {
  return new DynamoDbDataSource(api, "main", {
    api: api,
    table: props.ddb,
    name: "ddb",
    readOnlyAccess: true,
  });
};

class QueryResolverInput {
  stack: ApiStack;
  resolverNames: string[];
  target: environment.Environments;
  functionRole: IAM.PolicyStatement;
  constructor(
    stack: ApiStack,
    resolverNames: string[],
    target: environment.Environments,
    functionRole: IAM.PolicyStatement
  ) {
    this.stack = stack;
    this.resolverNames = resolverNames;
    this.target = target;
    this.functionRole = functionRole;
  }
}

const apiDefaultRole = (
  api: GraphqlApi,
  target: environment.Environments,
  props: ApiProps
) => {
  const serviceRole = new IAM.Role(api, "appsyncServiceRole", {
    assumedBy: new IAM.ServicePrincipal("appsync.amazonaws.com"),
    path: "/service-role/",
    roleName: environment.withEnvPrefix(target, "appsyncDefaultRole"),
  });
  return serviceRole;
};

const withLambdaQueryResolvers = (
  api: GraphqlApi,
  input: QueryResolverInput,
  target: environment.Environments
) => {
  input.resolverNames.forEach((n) => {
    const f = lambdaFunction(input.stack, n, target);
    f.addToRolePolicy(input.functionRole);
    api.addLambdaDataSource(n, f).createResolver(queryResolver(n));
  });
};

const resolver = (typeName: string, fieldName: string) => {
  return {
    typeName: typeName,
    fieldName: fieldName,
    responseMappingTemplate: MappingTemplate.lambdaResult(),
  };
};

const queryResolver = (fieldName: string) => {
  return resolver("Query", fieldName);
};

const apiProps = (target: environment.Environments): GraphqlApiProps => {
  return {
    xrayEnabled: true,
    name: environment.withEnvPrefix(target, "GraphqlApi"),
    logConfig: {
      fieldLogLevel: FieldLogLevel.ALL,
      excludeVerboseContent: false,
    },
    schema: Schema.fromAsset(join(__dirname, "schema/schema.graphql")),
    authorizationConfig: {
      defaultAuthorization: {
        authorizationType: AuthorizationType.API_KEY,
        apiKeyConfig: {
          expires: cdk.Expiration.after(cdk.Duration.days(365)),
          description: "default",
        },
      },
    },
  };
};
