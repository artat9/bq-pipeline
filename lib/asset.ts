import { ICertificate } from "@aws-cdk/aws-certificatemanager";
import {
  CloudFrontAllowedMethods,
  CloudFrontWebDistribution,
  OriginAccessIdentity,
  PriceClass,
  SSLMethod,
} from "@aws-cdk/aws-cloudfront";
import {
  CanonicalUserPrincipal,
  Effect,
  PolicyStatement,
} from "@aws-cdk/aws-iam";
import { S3EventSource } from "@aws-cdk/aws-lambda-event-sources";
import { ARecord, IHostedZone, RecordTarget } from "@aws-cdk/aws-route53";
import { CloudFrontTarget } from "@aws-cdk/aws-route53-targets";
import {
  BlockPublicAccess,
  Bucket,
  BucketAccessControl,
  EventType,
  HttpMethods,
} from "@aws-cdk/aws-s3";
import {
  Construct,
  Duration,
  RemovalPolicy,
  Stack,
  StackProps,
} from "@aws-cdk/core";
import * as environment from "./env";
import { lambdaFunction } from "./function";
type AsstesStackProps = StackProps & {
  certificate: ICertificate;
  hostedZone: IHostedZone;
};
export class AsstesStack extends Stack {
  public readonly assetBucket: Bucket;
  constructor(
    scope: Construct,
    id: string,
    target: environment.Environments,
    props: AsstesStackProps
  ) {
    super(scope, id);
    permanentBucket(this, target, this.account, props);
  }
}

export const assetBucketName = (target: environment.Environments) => {
  return environment.withEnvPrefix(target, "asset");
};

const permanentBucket = (
  scope: Construct,
  target: environment.Environments,
  account: string,
  props: AsstesStackProps
) => {
  const b = new Bucket(scope, "AssetBucket", {
    accessControl: BucketAccessControl.PRIVATE,
    blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
    bucketName: assetBucketName(target),
    publicReadAccess: false,
    removalPolicy: RemovalPolicy.DESTROY,
    autoDeleteObjects: true,
    cors: cors(),
  });
  const oai = new OriginAccessIdentity(scope, "OAI-AssetsBucket");
  const myBucketPolicy = new PolicyStatement({
    effect: Effect.ALLOW,
    actions: ["s3:GetObject"],
    principals: [
      new CanonicalUserPrincipal(
        oai.cloudFrontOriginAccessIdentityS3CanonicalUserId
      ),
    ],
    resources: [`${b.bucketArn}/*`],
  });
  b.addToResourcePolicy(myBucketPolicy);
  const assetsDist = new CloudFrontWebDistribution(scope, "AssetDistribution", {
    viewerCertificate: {
      aliases: [assetDomainName(target)],
      props: {
        acmCertificateArn: props.certificate.certificateArn,
        minimumProtocolVersion: "TLSv1",
        sslSupportMethod: SSLMethod.SNI,
      },
    },
    priceClass: PriceClass.PRICE_CLASS_ALL,
    originConfigs: [
      {
        s3OriginSource: {
          s3BucketSource: b,
          originAccessIdentity: oai,
        },
        behaviors: [
          {
            isDefaultBehavior: true,
            allowedMethods: CloudFrontAllowedMethods.GET_HEAD_OPTIONS,
            minTtl: Duration.days(365),
            maxTtl: Duration.days(365),
            defaultTtl: Duration.days(365),
          },
        ],
      },
    ],
  });
  const invalidateFunc = lambdaFunction(scope, "updateimage", target, {
    AssetDistributionId: assetsDist.distributionId,
  });
  invalidateFunc.addToRolePolicy(
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ["cloudfront:CreateInvalidation"],
      resources: [
        `arn:aws:cloudfront::${account}:distribution/${assetsDist.distributionId}`,
        `arn:aws:cloudfront::${account}:distribution/${assetsDist.distributionId}/*`,
      ],
    })
  );
  invalidateFunc.addEventSource(
    new S3EventSource(b, {
      events: [EventType.OBJECT_CREATED],
    })
  );
  new ARecord(scope, `AssetCDNArecord`, {
    recordName: assetDomainName(target),
    target: RecordTarget.fromAlias(new CloudFrontTarget(assetsDist)),
    zone: props.hostedZone,
  });
  return b;
};

const assetDomainName = (target: environment.Environments) => {
  const { rootDomain } = environment.valueOf(target);
  return `assets.${rootDomain}`;
};

const cors = () => {
  return [
    {
      allowedMethods: [HttpMethods.GET, HttpMethods.HEAD],
      allowedOrigins: ["*"],
      allowedHeaders: ["*"],
      maxAge: 3000,
      exposedHeaders: [
        "x-amz-server-side-encryption",
        "x-amz-request-id",
        "x-amz-id-2",
        "ETag",
      ],
    },
  ];
};
