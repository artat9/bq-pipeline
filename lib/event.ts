import { Table } from '@aws-cdk/aws-dynamodb';
import { IEventSource } from '@aws-cdk/aws-lambda';
import { SnsEventSource } from '@aws-cdk/aws-lambda-event-sources';
import { Topic } from '@aws-cdk/aws-sns';
import { Construct, Stack, StackProps } from '@aws-cdk/core';
import * as environment from './env';
import { lambdaFunction } from './function';

type EventProps = StackProps & {
  ddb: Table;
  esEndPoint: string;
  esArn: string;
  publicApiLogGroup: string;
};

export class EventStack extends Stack {
  constructor(
    scope: Construct,
    id: string,
    target: environment.Environments,
    props?: StackProps
  ) {
    super(scope, id, props);
    const applicationCreated = new Topic(this, 'application-created', {
      topicName: applicationCreatedTopicName(target),
    });
    subscribeFunctions(
      this,
      new SnsEventSource(applicationCreated),
      ['accountmail', 'notifytoslack'],
      target
    );
  }
}

const subscribeFunctions = (
  scope: Construct,
  source: IEventSource,
  functions: string[],
  target: environment.Environments
) => {
  functions.forEach((f) => {
    const func = lambdaFunction(scope, f, target);
    func.addEventSource(source);
  });
};

export const applicationCreatedTopicName = (
  target: environment.Environments
) => {
  return environment.withEnvPrefix(target, 'media-application-created');
};
