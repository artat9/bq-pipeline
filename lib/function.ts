import { Effect, PolicyStatement } from '@aws-cdk/aws-iam';
import { Code, Function, Runtime, Tracing } from '@aws-cdk/aws-lambda';
import { Construct, Duration } from '@aws-cdk/core';
import { resolve } from 'path';
import * as environment from './env';
import { applicationCreatedTopicName } from './event';

export const lambdaFunction = (
  scope: Construct,
  id: string,
  target: environment.Environments,
  additionalEnvParams?: { [key: string]: string }
) => {
  const func = new Function(scope, id, {
    functionName: environment.withEnvPrefix(target, id),
    code: code(id),
    handler: 'bin/main',
    timeout: Duration.minutes(1),
    runtime: Runtime.GO_1_X,
    tracing: Tracing.ACTIVE,
    environment: environmentParameters(target, additionalEnvParams),
  });
  // TODO: restrict access
  func.addToRolePolicy(
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ['lambda:*'],
      resources: ['*'],
    })
  );
  func.addToRolePolicy(
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ['ssm:Get*'],
      resources: ['*'],
    })
  );
  func.addToRolePolicy(
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ['dynamodb:*'],
      resources: ['*'],
    })
  );
  func.addToRolePolicy(
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ['sns:Publish'],
      resources: ['*'],
    })
  );
  func.addToRolePolicy(
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ['ses:SendEmail'],
      resources: ['*'],
    })
  );
  func.addToRolePolicy(
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ['s3:*'],
      resources: ['*'],
    })
  );
  return func;
};
const code = (dirname: string) => {
  return Code.fromAsset(
    resolve(`${__dirname}/../`, 'cmd', dirname, 'bin', 'main.zip')
  );
};

const environmentParameters = (
  target: environment.Environments,
  additionalEnvParams?: {
    [key: string]: string;
  }
): { [key: string]: string } => {
  var stack: { [key: string]: string } = {
    EnvironmentId: target.toString(),
    AllowedOrigin: environment.valueOf(target).allowedOrigin,
    ApplicationCreatedTopicName: applicationCreatedTopicName(target),
    RootDomain: environment.valueOf(target).rootDomain,
  };
  if (!additionalEnvParams) {
    return stack;
  }
  Object.keys(additionalEnvParams).forEach((k) => {
    stack[k] = additionalEnvParams[k];
  });
  return stack;
};
