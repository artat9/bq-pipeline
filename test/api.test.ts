import { SynthUtils } from "@aws-cdk/assert";
import { App } from "@aws-cdk/core";
import { Environments } from "../lib/env";
import { ApiStack } from "./../lib/api";
import { DataSourceStack } from "./../lib/datasource";

test("resource created", () => {
  const app = new App();
  const target = Environments.TEST;
  const ds = new DataSourceStack(app, "ds", target);
  const stack = new ApiStack(app, "test", target, {
    ddb: ds.ddb,
  });
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
