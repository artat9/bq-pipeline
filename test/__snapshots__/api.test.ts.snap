// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resource created 1`] = `
Object {
  "Resources": Object {
    "GraphqlApi1B6CF24C": Object {
      "Properties": Object {
        "AuthenticationType": "API_KEY",
        "LogConfig": Object {
          "CloudWatchLogsRoleArn": Object {
            "Fn::GetAtt": Array [
              "GraphqlApiApiLogsRoleBB9E6BAD",
              "Arn",
            ],
          },
          "ExcludeVerboseContent": false,
          "FieldLogLevel": "ALL",
        },
        "Name": "kaleido-backend-GraphqlApi-test",
        "XrayEnabled": true,
      },
      "Type": "AWS::AppSync::GraphQLApi",
    },
    "GraphqlApiApiLogsRoleBB9E6BAD": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "appsync.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSAppSyncPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GraphqlApiDefaultApiKey47EE7128": Object {
      "DependsOn": Array [
        "GraphqlApiSchema1B71CE4F",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GraphqlApi1B6CF24C",
            "ApiId",
          ],
        },
        "Description": "default",
        "Expires": 1666489880,
      },
      "Type": "AWS::AppSync::ApiKey",
    },
    "GraphqlApiQuerygetAccountResolver244E426B": Object {
      "DependsOn": Array [
        "GraphqlApimain6079C13B",
        "GraphqlApiSchema1B71CE4F",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GraphqlApi1B6CF24C",
            "ApiId",
          ],
        },
        "DataSourceName": "ddb",
        "FieldName": "getAccount",
        "Kind": "UNIT",
        "RequestMappingTemplate": "{
    "version" : "2017-02-28",
    "operation" : "Query",
    "query" : {
        ## Provide a query expression. **
        "expression": "PK = :PK",
        "expressionValues" : {
            ":PK" : $util.dynamodb.toDynamoDBJson("Account:\${util.str.toLower(\${context.arguments.address})}")
        }
    },
}
",
        "ResponseMappingTemplate": "
#set($address = "")
#set($nonce="")
#if ($ctx.result.items.size() == 0)
    #return
#end
#foreach($item in $ctx.result.items)
    #set($address =  $util.str.toReplace($item.PK, "Account:", ""))
    #set($nonce =  $item.Data)
#end
{
  "address": $util.toJson($address),
  "nonce": $util.toJson($nonce)
}
",
        "TypeName": "Query",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "GraphqlApiSchema1B71CE4F": Object {
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GraphqlApi1B6CF24C",
            "ApiId",
          ],
        },
        "Definition": "type Query {
  getAccount(address: String!): Account
}

type Account {
  address: String!
  nonce: String!
  iconPath: String
  bannerPath: String
}

",
      },
      "Type": "AWS::AppSync::GraphQLSchema",
    },
    "GraphqlApimain6079C13B": Object {
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GraphqlApi1B6CF24C",
            "ApiId",
          ],
        },
        "DynamoDBConfig": Object {
          "AwsRegion": Object {
            "Ref": "AWS::Region",
          },
          "TableName": Object {
            "Fn::ImportValue": "ds:ExportsOutputRefkaleidobackendmaintest9352CC46D953DB2A",
          },
        },
        "Name": "ddb",
        "ServiceRoleArn": Object {
          "Fn::GetAtt": Array [
            "GraphqlApimainServiceRole47381CDD",
            "Arn",
          ],
        },
        "Type": "AMAZON_DYNAMODB",
      },
      "Type": "AWS::AppSync::DataSource",
    },
    "GraphqlApimainServiceRole47381CDD": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "appsync.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "GraphqlApimainServiceRoleDefaultPolicy24D667BC": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::ImportValue": "ds:ExportsOutputFnGetAttkaleidobackendmaintest9352CC46Arn7DF614B1",
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::ImportValue": "ds:ExportsOutputFnGetAttkaleidobackendmaintest9352CC46Arn7DF614B1",
                      },
                      "/index/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GraphqlApimainServiceRoleDefaultPolicy24D667BC",
        "Roles": Array [
          Object {
            "Ref": "GraphqlApimainServiceRole47381CDD",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
